File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    vector<vector<int>>dirs={{1,0},{0,1},{-1,0},{0,-1}};
    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int n=matrix.size();
        int m=matrix[0].size();
        int longestPath=0;
        
        vector<vector<int>>dp(n,vector<int>(m));
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                longestPath=max(longestPath,dfs(matrix,i,j,n,m,dp));
            }
        }
        
        return longestPath;
    }
    
    int dfs(vector<vector<int>>& matrix,int r,int c,int n,int m,vector<vector<int>>&dp){
        int maxPath=0;
        
        if(dp[r][c]>0){
            return dp[r][c];
        }
        
        for(auto dir:dirs){
            int rr=r+dir[0];
            int cc=c+dir[1];
            
            if(rr>=0 && cc>=0 && rr<n && cc<m && matrix[rr][cc]>matrix[r][c]){
               maxPath=max(maxPath,dfs(matrix,rr,cc,n,m,dp));
            }
        }
        
        return dp[r][c]=maxPath+1;
    }
};

File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.java
class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        int n=matrix.length;
        int m=matrix[0].length;
        
        int[][]dir={{1,0},{0,1},{0,-1},{-1,0}};
        
        int[][]indegree=new int[n][m];
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                
                for(int d=0;d<dir.length;d++){
                    int r=dir[d][0]+i;
                    int c=dir[d][1]+j;
                    
                    if(r>=0 && c>=0 && r<n && c<m && matrix[r][c]<matrix[i][j]){
                        indegree[r][c]++;
                    }
                }
            }
        }
        
        LinkedList<Integer>q=new LinkedList<>();
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(indegree[i][j]==0){
                    q.addLast(i*m+j);
                }
            }
        }
        
        int level=0;
        
        while(q.size()>0){
            int size=q.size();
            while(size-->0){
                int vtx=q.removeFirst();
                int rr=vtx/m;
                int cc=vtx%m;
                
                for(int i=0;i<dir.length;i++){
                    int r=rr+dir[i][0];
                    int c=cc+dir[i][1];
                    
                    if(r>=0 && c>=0 && r<n && c<m && matrix[r][c] < matrix[rr][cc]){
                        indegree[r][c]--;
                        if(indegree[r][c]==0){
                            q.addLast(r*m+c);
                        }
                    }
                }
                
            }
            level++;
        }
        
        return level;
    }
}